enhancements/repositories-auto-pruning.md

```diff
@@ -0,0 +1,179 @@
+{
+    fieldToBeUpdated: value,
+}
+```
+
+**List policies**
+```
+GET /api/v1/repository/{apirepopath:repository}/autoprunepolicy
+```
+
+**Get policy**
+```
+GET /api/v1/repository/{apirepopath:repository}/autoprunepolicy/{policyid}
+```
+
+**Delete policy**
+```
+DELETE /api/v1/repository/{apirepopath:repository}/autoprunepolicy/{policyid}
+```
+
+### Database structure
+A new table will be created to store repository policies.
+
+**repositoryautoprunepolicy**
+
+This table holds the policy configuration for a single repository. For this proposal there will be only one entry per repository but there will be support for multiple rows per `repository_id` in future. The policy field will hold the policy details such as `{method: "creation_date", olderThan: "2w"}` or `{method: "number_of_tags",  numTags: 100}`. `repository_id` is a foreign key that references the `repositories` table.
+
+| Field | Datatype | Attributes | Description|
+| --- | ----------- | ----------- | ----------- |
+| uuid | character varying(255) | Unique, not null, indexed | Unique identifier for this policy |
+| repository_id | integer | FK, not null | Repository that the policy falls under |
+| policy | text | JSON, not null | Policy configuration |
+
+**autoprunetaskstatus**
+
+This existing table will continue to be used to execute repository pruning policies. A new entry will only be created if there is no entry for the repo's namespace for a given repository policy. 
+
+### Auto-prune worker
+
+**Creation of an auto-prune task**
+
+Whenever a policy is created in the `repositoryautoprunepolicy` table, there can be two cases: 
+* if the `autoprunetaskstatus` doesn't have a row for the namespace the repository belongs to - a new row is created with the repo's namespace as `namespace_id`, `status="queued"` and `last_ran_ms=None` in the `autoprunetask` table. This will be done within the same transaction. 
+* if the `autoprunetaskstatus` already has a row for the namespace - no update is done to the `autoprunetaskstatus` table.
+
+The auto-prune worker will use the entry in the `autoprunetaskstatus` table to identify which namespaces it should execute policies for.
+
+**Execution of the auto-prune worker**
+
+The existing auto-pruning worker will be refactored to support both namespace and repository level policies. It's workflow is based on values in the `autoprunetaskstatus` table.
+
+**Worker Pseudocode**
+
+* Worker starts on a set interval (1hr)
+* Select a row from `autoprunetaskstatus` with the least or null `last_ran_ms` and `FOR UPDATE SKIP LOCKED`
+    * A null `last_ran_ms` indicates that the task was never ran
+    * A task that hasn't been ran in the longest amount of time or has never been ran at all are prioritized.
+* Get the namespace policy(s) configuration for the namespace from `namespaceautoprunepolicy`
+    * If no namespace policy exists, look up repositories under the namespace containing repository policy(s). 
```

Comment:
here we only need to check that one repository policy exists. Could probably use peewee's `.exists()`